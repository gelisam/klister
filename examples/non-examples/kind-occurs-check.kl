#lang "prelude.kl"

(datatype (Phantom A)
  (mk-phantom))

-- causes the kind equality ?1 ~ ?2 to be unified twice, which previously
-- caused ?2 to point to itself, thus causing zonking ?2 to run forever.
-- should be accepted.
(example
  (with-unknown-type [A]
    (the (Phantom A)
      (mk-phantom))))

-- causes the kind equality ?1 ~ (-> ?1 *).
-- should be accepted by the occurs-check.
(datatype (InfiniteKind A)
  (mk-infinite-kind (Phantom (A A))))
