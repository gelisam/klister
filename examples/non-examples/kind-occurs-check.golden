Infinite kind detected:
  (KindVar 55708) = (META(KindVar 55708) â†’ META(KindVar 55701))
